### ISSUE TITLE
Add ETW events that would measure application start up time.

### ISSUE THREAD

**Issue by Tanya-Solyanik** [dotnet/winforms#6646]
This is a request from the runtime team, aligned with their Arm64 effort. Related to https://github.com/dotnet/runtime/issues/64993

As an example, here’s a pointer to how the perf lab does this.  All of the start-up tests emit an event when they have completed start-up: [https://github.com/dotnet/performance/blob/23b05253055e3c5253bf4a1c8d4d767f33c2b77b/src/tools/ScenarioMeasurement/Startup/GenericStartupParser.cs#L8](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fperformance%2Fblob%2F23b05253055e3c5253bf4a1c8d4d767f33c2b77b%2Fsrc%2Ftools%2FScenarioMeasurement%2FStartup%2FGenericStartupParser.cs%23L8&data=04%7C01%7CTatiana.Solyanik%40microsoft.com%7Cc3e0644499224ac8f73c08d9eb2a047e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637799387966396924%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=pLz5axwT4EfpYxtAjsYiGIBarg43zoIXzRrvXOx3mdM%3D&reserved=0)

**Comment by merriemcgaw** [comment]
Definitely something we want to try to measure.

**Comment by dreddy-work** [comment]
@Tanya-Solyanik, Are you planning to take this up in 7.0?

**Comment by Tanya-Solyanik** [comment]
>@Tanya-Solyanik, Are you planning to take this up in 7.0?

No, this is a request from the runtime team. Someone from our team should work with Rich to understand his goal, and then we need to decide if his goals align with ours.

**Comment by elachlan** [comment]
This would be nice to have. One of my main goals for working on Winforms was to try and speed up app performance. We currently rely heavily on NGEN as a part of the install process. Currently .NET 6 starts significantly slower.

Related runtime issue which has an example app: https://github.com/dotnet/runtime/issues/13339

**Comment by JeremyKuhne** [comment]
Measuring time isn't really generically practical as there aren't great places to measure and we're at the mercy of Windows message processing. When I was working on startup performance I was seeing wildly fluctuating times on `Application.Run` for a new `Form()`.

What we do need and should have are allocation tracking tests for startup scenarios. We can do a hacky bit where we implement some tests that expect a specific amount of thread allocations, but they'll need a lot of attention as changes in the lower level runtime will also impact this from time to time. It would be best to get a perf test solution up and running but I don't see that we'll get that scheduled in the near future.

@elachlan if you can come up with measurement points that you think would be useful we can discuss.

**Comment by elachlan** [comment]
Thanks for the information, I guess the ETW events would help application developers profile their code better in production scenarios.

For myself, I am interested in general performance improvements across the board. Not just start-up performance.

**Comment by jkotas** [comment]
> What we do need and should have are allocation tracking tests for startup scenarios. 

Note that the common techniques for reducing allocations often regress startup performance. GC allocations are not a good proxy for startup performance. The best proxy for tracking startup performance of .NET apps is total number and code size of managed methods that are touched during startup.
